### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by tcharding**
Draft for initial feedback please. (#498 is the same but uses `log` crate)

- Add logging infrastructure as described in #495, logs to a file hardcode as a string in the `logging` module.
- Add logging when signing.
- Add logging when creating private keys.

Fix: #495

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,9 @@ global-context = ["std"]
 # and is not necessary.)
 global-context-less-secure = ["global-context"]
 
+# Enables debug logging, do not enable in production systems.
+danger_leak_secret_material = ["std"]
```
Comment (apoelstra) [inline Cargo.toml]: In a84629c3cff592cf1799b50fb60ea8e052c565fa and 8cb37e093bf91bd1f621af5f35ca473153bd5884:

As discussed elsewhere we cannot use a cargo feature for this.

File: src/logging.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#[allow(dead_code)] // When danger_leak_secret_material not enabled.
+pub static LOGFILE: Mutex<Option<File>> = Mutex::new(None);
+
+#[allow(dead_code)] // When danger_leak_secret_material not enabled.
+pub const CRYPTO_DEBUG_FILE: &str = "/tmp/rust-secp256k1.log";
+
+/// Logs a format string to `CRYPTO_DEBUG_FILE`.
+#[cfg(feature = "danger_leak_secret_material")]
+#[macro_export]
+macro_rules! log {
```
Comment (apoelstra) [inline src/logging.rs:20]: In 8cb37e093bf91bd1f621af5f35ca473153bd5884:

This should have a different name to make it clearer that it's a debug-only thing. Maybe `trace_log` or just `trace`.

**Comment by apoelstra** [comment]
It really seems like the `log` crate isn't getting us much here, so we ought to just do the homebrew thing.

I think we should use `{}` rather than `{:?}`.

Does the "`KeyPair` overflows" thing happen on both PRs or just in #498 ?

File: Cargo.toml (hunk 1 above)
Comment (tcharding) [reply Cargo.toml]: My bad, I had not worked on this since all our chats. Sorry for not making that explicit and wasting your review time.

**Comment by tcharding** [comment]
> Does the "KeyPair overflows" thing happen on both PRs or just in https://github.com/rust-bitcoin/rust-secp256k1/pull/498 ?

I only noticed it in #498 but I'll see as I play more with this one, will keep you posted.

File: src/logging.rs (hunk 2 above)
Comment (tcharding) [reply src/logging.rs:20]: Could still be confused with `trace` macro: https://docs.rs/log/latest/log/macro.trace.html. What about `leak!` if we use the `pub(crate)` way of exporting the macro usage could be `logging::leak!("...")`?

Comment (tcharding) [reply src/logging.rs:20]: I'll play with both and see how they feel. Cheers.

**Comment by tcharding** [comment]
I don't see myself getting back to this any time soon.

**Pull request closed without merging**